### PR TITLE
ClassLoader/ModelFileDAO fixes

### PR DESCRIPTION
**Pull request by mdittmer**
This was work towards fixing up ClassLoader/ModelFileDAO before #262 was identified.

(1) account for require() only running code the first time;
(2) validate models at the appropriate time relative to requires axioms;
(3) check model ids before returning them from find().

### CODE REVIEW

File: src/foam/classloader/ClassLoader.js [hunk 1]
```diff
@@ -82,8 +82,11 @@ foam.CLASS({
 
         var modelDao = X[foam.String.daoize(foam.core.Model.name)];
         this.pending[modelId] = modelDao.find(modelId).then(function(m) {
-          m.validate();
-          return m.arequire(deps).then(function() { return m; });
+          modelId;
```
Comment (adamvy) [inline src/foam/classloader/ClassLoader.js]: Why are you just referencing modelId here?

File: src/foam/classloader/NodeModelFileDAO.js [hunk 2]
```diff
@@ -23,20 +23,40 @@ foam.CLASS({
   properties: [
     {
       name: 'classpath'
+    },
+    {
+      name: 'loadedModels',
```
Comment (adamvy) [inline src/foam/classloader/NodeModelFileDAO.js]: Mike had this support originally but I had him remove it.  It's not the ModelDAO's job to cache IMO.

File: src/foam/classloader/WebModelFileDAO.js [hunk 3]
```diff
@@ -52,10 +52,19 @@ foam.CLASS({
         var model;
         var foamCLASS = foam.CLASS;
 
-        foam.CLASS = function(m) {
-          var cls = m.class ? foam.lookup(m.class) : foam.core.Model;
-          model = cls.create(m, self);
-          foam.CLASS = foamCLASS;
+        foam.CLASS = function(clsDef) {
+          var classOfModel = clsDef.class ? foam.lookup(clsDef.class) :
```
Comment (adamvy) [inline src/foam/classloader/WebModelFileDAO.js]: I do not find this wordier version easier to read.  In fact I find it less readable.

File: src/foam/classloader/ClassLoader.js (hunk 1 above)
Comment (mdittmer) [reply src/foam/classloader/ClassLoader.js]: Leftover from debugging. Will fix.

File: src/foam/classloader/NodeModelFileDAO.js (hunk 2 above)
Comment (mdittmer) [reply src/foam/classloader/NodeModelFileDAO.js]: Who tracks whether code has already been `require`d, then?

File: src/foam/classloader/WebModelFileDAO.js (hunk 3 above)
Comment (mdittmer) [reply src/foam/classloader/WebModelFileDAO.js]: It's wordier, but it describes the meaning of the model/class back-flips we're doing here.

@jacksonic @mcarcaso @shepheb @kgrgreer am I the only one that finds this easier to parse? If so, I'll happily change it back, and maybe add a short comment instead.

File: src/foam/classloader/NodeModelFileDAO.js (hunk 2 above)
Comment (adamvy) [reply src/foam/classloader/NodeModelFileDAO.js]: Context.arequire() does the check.

https://github.com/foam-framework/foam2/blob/master/src/foam/classloader/ClassLoader.js#L78

It's actually more complicated than just caching the loaded models.  Cause we also do special bookkeeping to resolve circular dependencies.

File: src/foam/classloader/WebModelFileDAO.js (hunk 3 above)
Comment (adamvy) [reply src/foam/classloader/WebModelFileDAO.js]: classOfModel followed by modelOfClass is confusing, because the modelOfClass is not the model of the class referenced in classOfModel.

The confusing bit here (IMO) is that we're overriding foam.CLASS() to extract the model definition from a file which contains a foam.CLASS call.   These descriptive names do not adequately describe the "backflips" we're doing, they just make it longer which makes it even harder to parse IMO.   I would rather keep the code as is and add a small comment explaining what we're doing.

Once you understand what this code is trying to achieve I think it reads fine as is.

File: src/foam/classloader/NodeModelFileDAO.js (hunk 2 above)
Comment (mdittmer) [reply src/foam/classloader/NodeModelFileDAO.js]: That's contextualized. `require()` is global. Re-`arequire`-ing in different contexts is what instigated this change.

Comment (adamvy) [reply src/foam/classloader/NodeModelFileDAO.js]: Ah I see.

Maybe require wasn't the best thing to use here, but rather readFile and eval/vm.runInThisContext

But I guess this is okay too.

File: src/foam/classloader/WebModelFileDAO.js (hunk 3 above)
Comment (mdittmer) [reply src/foam/classloader/WebModelFileDAO.js]: You've convinced me. PTAL at comments added. The only "rename" change I kept is not double-declaring `cls`. I think using the same variable to mean something entirely different was what instigated the confusion for me.

Comment (mdittmer) [reply src/foam/classloader/WebModelFileDAO.js]: Oops. New changes were added to #259. PTAL there. I'll make the code here match.

File: src/foam/classloader/NodeModelFileDAO.js (hunk 2 above)
Comment (mdittmer) [reply src/foam/classloader/NodeModelFileDAO.js]: I would generally agree. Shall I add a TODO?

File: src/foam/classloader/WebModelFileDAO.js (hunk 3 above)
Comment (mdittmer) [reply src/foam/classloader/WebModelFileDAO.js]: Now updated here.

File: src/foam/classloader/NodeModelFileDAO.js (hunk 2 above)
Comment (mdittmer) [reply src/foam/classloader/NodeModelFileDAO.js]: nvm; just changed it. PTAL.

File: src/foam/classloader/ClassLoader.js [hunk 4]
```diff
@@ -82,8 +82,10 @@ foam.CLASS({
 
         var modelDao = X[foam.String.daoize(foam.core.Model.name)];
         this.pending[modelId] = modelDao.find(modelId).then(function(m) {
-          m.validate();
-          return m.arequire(deps).then(function() { return m; });
+          return m.arequire(deps).then(function() {
+            m.validate();
```
Comment (kgrgreer) [inline src/foam/classloader/ClassLoader.js]: Add comment to explain why this is done.

File: src/foam/classloader/NodeModelFileDAO.js [hunk 5]
```diff
@@ -21,36 +21,51 @@ foam.CLASS({
...
+    { name: 'fs',  factory: function() { return require('fs');       } },
+    { name: 'vm',  factory: function() { return require('vm');       } }
   ],
 
   methods: [
     function find(id) {
       var foamCLASS = foam.CLASS;
       var self = this;
-      var model;
+      var model = null;
```
Comment (kgrgreer) [inline src/foam/classloader/NodeModelFileDAO.js]: Why is this change needed?

File: src/foam/classloader/NodeModelFileDAO.js [hunk 6]
```diff
@@ -21,36 +21,51 @@ foam.CLASS({
...
-        return Promise.resolve(null);
-      } finally {
-        foam.CLASS = foamCLASS;
-      }
-
-      return Promise.resolve(model);
+        var cls = m.class ? foam.lookup(m.class) :
+              foam.core.Model;
+        var mdl = cls.create(m, self);
+        if ( mdl.id === id ) {
```
Comment (kgrgreer) [inline src/foam/classloader/NodeModelFileDAO.js]: When would mdl.id !== id?

File: src/foam/classloader/WebModelFileDAO.js [hunk 7]
```diff
@@ -53,9 +53,18 @@ foam.CLASS({
         var foamCLASS = foam.CLASS;
 
         foam.CLASS = function(m) {
-          var cls = m.class ? foam.lookup(m.class) : foam.core.Model;
-          model = cls.create(m, self);
-          foam.CLASS = foamCLASS;
+          var cls = m.class ? foam.lookup(m.class) :
+                foam.core.Model;
+          var mdl = cls.create(m, self);
+          if ( mdl.id === id ) {
```
Comment (kgrgreer) [inline src/foam/classloader/WebModelFileDAO.js]: When is this not the case?

File: src/foam/classloader/NodeModelFileDAO.js (hunk 5 above)
Comment (mdittmer) [reply src/foam/classloader/NodeModelFileDAO.js]: The DAO should `resolve(null)` if the required code doesn't actually provide the requested class model.

File: src/foam/classloader/NodeModelFileDAO.js (hunk 6 above)
Comment (mdittmer) [reply src/foam/classloader/NodeModelFileDAO.js]: If the file name matches, but it contains multiple `CLASS` calls (some of which may not be of the requested id).

File: src/foam/classloader/WebModelFileDAO.js (hunk 7 above)
Comment (mdittmer) [reply src/foam/classloader/WebModelFileDAO.js]: If the file name matches, but it contains multiple `CLASS` calls (some of which may not be of the requested id).

File: src/foam/classloader/NodeModelFileDAO.js (hunk 6 above)
Comment (kgrgreer) [reply src/foam/classloader/NodeModelFileDAO.js]: Add a comment to explain that.

**Comment by mdittmer** [comment]
@kgrgreer @adamvy this is getting stale. Anything else before we merge it in?

**Comment by kgrgreer** [comment]
LGTM
